pyomo/contrib/incidence_analysis/scc_solver.py

```diff
@@ -0,0 +1,169 @@
+    ]
+    for block, inputs in generate_subsystem_blocks(
+        subsets,
+        include_fixed=include_fixed,
+    ):
+        # TODO: How does len scale for reference-to-list?
+        assert len(block.vars) == len(block.cons)
+        yield (block, inputs)
+
+
+def solve_strongly_connected_components(
+    block,
+    solver=None,
+    solve_kwds=None,
+    calc_var_kwds=None,
+):
+    """Solve a square system of variables and equality constraints by
+    solving strongly connected components individually.
+
+    Strongly connected components (of the directed graph of constraints
+    obtained from a perfect matching of variables and constraints) are
+    the diagonal blocks in a block triangularization of the incidence
+    matrix, so solving the strongly connected components in topological
+    order is sufficient to solve the entire block.
+
+    One-by-one blocks are solved using Pyomo's
+    calculate_variable_from_constraint function, while higher-dimension
+    blocks are solved using the user-provided solver object.
+
+    Parameters
+    ----------
+    block: Pyomo Block
+        The Pyomo block whose variables and constraints will be solved
+    solver: Pyomo solver object
+        The solver object that will be used to solve strongly connected
+        components of size greater than one constraint. Must implement
+        a solve method.
+    solve_kwds: Dictionary
+        Keyword arguments for the solver's solve method
+    calc_var_kwds: Dictionary
+        Keyword arguments for calculate_variable_from_constraint
+
+    Returns
+    -------
+    List of results objects returned by each call to solve
+
+    """
+    if solve_kwds is None:
+        solve_kwds = {}
+    if calc_var_kwds is None:
+        calc_var_kwds = {}
+
+    constraints = list(block.component_data_objects(Constraint, active=True))
+    var_set = ComponentSet()
+    variables = []
+    for con in constraints:
+        for var in identify_variables(con.expr, include_fixed=False):
+            # Because we are solving, we do not want to include fixed variables
+            if var not in var_set:
+                variables.append(var)
+                var_set.add(var)
+
+    res_list = []
+    for scc, inputs in generate_strongly_connected_components(
+        constraints,
+        variables,
+    ):
+        with TemporarySubsystemManager(to_fix=inputs):
+            if len(scc.vars) == 1:
+                _log.info(f"Solving 1x1 block: {scc.cons[0].name}.")
+                results = calculate_variable_from_constraint(
+                    scc.vars[0], scc.cons[0], **calc_var_kwds
+                )
+                res_list.append(results)
+            else:
+                if solver is None:
+                    # NOTE: Use local name to avoid slow generation of this
+                    # error message if a user provides a large, non-decomposable
+                    # block with no solver.
```

Comment:
Is it worth changing this to output the length of the lists, but truncate the names to the first *n* (5 or 10)?